stream.go

```diff
@@ -0,0 +1,331 @@
+package psrpc
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/livekit/psrpc/internal"
+)
+
+type Stream[SendType, RecvType proto.Message] interface {
+	Channel() <-chan *Response[RecvType]
+	Send(SendType) error
+	Close(error) error
+	Err() error
+}
+
+type ServerStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+	Hijack()
+}
+
+type ClientStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+}
+
+type streamAdapter interface {
+	send(ctx context.Context, msg *internal.Stream) error
+	close(streamID string)
+}
+
+type clientStream struct {
+	c     *RPCClient
+	rpc   string
+	topic string
+}
+
+func (s *clientStream) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.c.bus.Publish(ctx, getStreamServerChannel(s.c.serviceName, s.rpc, s.topic), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *clientStream) close(streamID string) {
+	s.c.mu.Lock()
+	delete(s.c.streamChannels, streamID)
+	s.c.mu.Unlock()
+}
+
+type serverStream[SendType, RecvType proto.Message] struct {
+	h      *streamRPCHandlerImpl[SendType, RecvType]
+	s      *RPCServer
+	nodeID string
+}
+
+func (s *serverStream[RequestType, ResponseType]) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.s.bus.Publish(ctx, getStreamChannel(s.s.serviceName, s.nodeID), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *serverStream[RequestType, ResponseType]) close(streamID string) {
+	s.h.mu.Lock()
+	delete(s.h.streams, streamID)
+	s.h.mu.Unlock()
+	s.h.handling.Dec()
+}
+
+type streamHandler[SendType, RecvType proto.Message] struct {
+	*streamImpl[SendType, RecvType]
+}
+
+func (h *streamHandler[SendType, RecvType]) Recv(msg proto.Message, err error) {
+	h.recv(msg, err)
```

Comment:
why duplicate these 3?

Reply:
`streamHandler` implements the `StreamHandler` interface returned by `StreamInterceptor`s. this is used as the last step in the interceptor chain to call `streamImpl`